### PR TITLE
Fix Plugin panic on restart

### PR DESCRIPTION
**Pull request by cpuguy83**
Fix panic on daemon restart with running plugin

Scenario:

Daemon is ungracefully shutdown and leaves plugins running (no
live-restore).
Daemon comes back up.
The next time a container tries to use that plugin it will cause a
daemon panic because the plugin client is not set.

This fixes that by ensuring that the plugin does get shutdown.
Note, I do not think there would be any harm in just re-attaching to the
running plugin instead of shutting it down, however historically we shut
down plugins and containers when live-restore is not enabled.

Note while attempting to fix this I ran into https://github.com/containerd/containerd/issues/2299 which required another workaround to prevent a double close on the plugin's exit chan.

### CODE REVIEW

File: plugin/executor/containerd/containerd.go [hunk 1]
```diff
@@ -98,23 +98,23 @@ func (e *Executor) Create(id string, spec specs.Spec, stdout, stderr io.WriteClo
...
 	alive, _, err := e.client.Restore(context.Background(), id, attachStreamsFunc(stdout, stderr))
 	if err != nil && !errdefs.IsNotFound(err) {
-		return err
+		return false, err
 	}
 	if !alive {
 		_, _, err = e.client.DeleteTask(context.Background(), id)
 		if err != nil && !errdefs.IsNotFound(err) {
-			logrus.WithError(err).Errorf("failed to delete container plugin %s task from containerd", id)
+			logrus.WithError(err).WithField("id", id).Errorf("failed to delete plugin container task from containerd")
```
Comment (anusha-ragunathan) [inline plugin/executor/containerd/containerd.go]: `plugin container task` is a bit much :) Can we stick to `plugin task`?

File: plugin/executor/containerd/containerd.go [hunk 2]
```diff
@@ -98,23 +98,23 @@ func (e *Executor) Create(id string, spec specs.Spec, stdout, stderr io.WriteClo
...
 		}
 
 		err = e.client.Delete(context.Background(), id)
 		if err != nil && !errdefs.IsNotFound(err) {
-			logrus.WithError(err).Errorf("failed to delete container plugin %s from containerd", id)
+			logrus.WithError(err).WithField("id", id).Errorf("failed to delete plugin container from containerd")
```
Comment (anusha-ragunathan) [inline plugin/executor/containerd/containerd.go]: Similarly, change message to `failed to delete plugin from containerd`

Comment (anusha-ragunathan) [reply plugin/executor/containerd/containerd.go]: Ditto for L138, L142

File: plugin/manager.go [hunk 3]
```diff
@@ -37,14 +37,14 @@ var validFullID = regexp.MustCompile(`^([a-f0-9]{64})$`)
 // Executor is the interface that the plugin manager uses to interact with for starting/stopping plugins
 type Executor interface {
 	Create(id string, spec specs.Spec, stdout, stderr io.WriteCloser) error
-	Restore(id string, stdout, stderr io.WriteCloser) error
 	IsRunning(id string) (bool, error)
+	Restore(id string, stdout, stderr io.WriteCloser) (running bool, err error)
```
Comment (anusha-ragunathan) [inline plugin/manager.go]: nit: `running` -> `alive` to be consistent.

File: plugin/manager.go [hunk 4]
```diff
@@ -138,17 +138,21 @@ func (pm *Manager) tmpDir() string {
 // HandleExitEvent is called when the executor receives the exit event
 // In the future we may change this, but for now all we care about is the exit event.
 func (pm *Manager) HandleExitEvent(id string) error {
+	logrus.WithField("id", id).Debug("HandleExitEvent")
```
Comment (anusha-ragunathan) [inline plugin/manager.go]: Remove.

File: plugin/manager.go [hunk 5]
```diff
@@ -138,17 +138,21 @@ func (pm *Manager) tmpDir() string {
...
 	p, err := pm.config.Store.GetV2Plugin(id)
 	if err != nil {
 		return err
 	}
 
-	os.RemoveAll(filepath.Join(pm.config.ExecRoot, id))
+	if err := os.RemoveAll(filepath.Join(pm.config.ExecRoot, id)); err != nil && !os.IsNotExist(err) {
+		logrus.WithError(err).WithField("plugin", id).Error("Could not remove plugin bundle dir")
```
Comment (anusha-ragunathan) [inline plugin/manager.go]: Lets pick one format for `WithField`. Either `WithField("plugin", ...)` like here or `WithField("id, ...")` like in `containerd.go`. I like the former

Comment (anusha-ragunathan) [reply plugin/manager.go]: Easier to grep in log files for the same field string.

File: plugin/manager.go [hunk 6]
```diff
@@ -205,11 +209,14 @@ func (pm *Manager) reload() error { // todo: restore
...
+		c := &controller{exitChan: make(chan bool)}
+		pm.mu.Lock()
 		pm.cMap[p] = c
+		pm.mu.Unlock()
+
 		go func(p *v2.Plugin) {
 			defer wg.Done()
-			if err := pm.restorePlugin(p); err != nil {
+			if err := pm.restorePlugin(p, c); err != nil {
 				logrus.Errorf("failed to restore plugin '%s': %s", p.Name(), err)
```
Comment (anusha-ragunathan) [inline plugin/manager.go]: Add `WithField` here?

File: plugin/manager_linux.go [hunk 7]
```diff
@@ -70,11 +70,12 @@ func (pm *Manager) enable(p *v2.Plugin, c *controller, force bool) error {
 }
 
 func (pm *Manager) pluginPostStart(p *v2.Plugin, c *controller) error {
+	logrus.WithField("plugin", p.Name()).Debug("plugin post start")
```
Comment (anusha-ragunathan) [inline plugin/manager_linux.go]: Remove this log. I'm assuming you added it for dev purposes. Dont see any value exposing this in the field.

**Comment by anusha-ragunathan** [comment]
`TestDaemonKillWithPlugins` is going to fail. That test expects the plugin to be up and running. You should update it.

File: plugin/executor/containerd/containerd.go (hunk 1 above)
Comment (cpuguy83) [reply plugin/executor/containerd/containerd.go]: Well, there's the task and the actual container obkect. Which both have to be dealt with.

File: plugin/manager.go (hunk 4 above)
Comment (cpuguy83) [reply plugin/manager.go]: Why?

File: plugin/executor/containerd/containerd.go (hunk 1 above)
Comment (anusha-ragunathan) [reply plugin/executor/containerd/containerd.go]: It may be better to add "plugin" in Withfield and remove it from the actual error message then?

File: plugin/manager.go (hunk 4 above)
Comment (anusha-ragunathan) [reply plugin/manager.go]: I assume it was added for development purposes only. Didnt see a real value add when in the field.

File: daemon/start.go [hunk 8]
```diff
@@ -176,19 +176,35 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint
 		return err
 	}
 
-	err = daemon.containerd.Create(context.Background(), container.ID, spec, createOptions)
+	ctx := context.TODO()
```
Comment (anusha-ragunathan) [inline daemon/start.go]: For my understanding: Why TODO context vs background?

File: daemon/start.go [hunk 9]
```diff
@@ -176,19 +176,35 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint
...
+	for i := 0; i < 2; i++ {
+		err = daemon.containerd.Create(context.Background(), container.ID, spec, createOptions)
```
Comment (anusha-ragunathan) [inline daemon/start.go]: sleep a tiny bit before retry?

Comment (cpuguy83) [reply daemon/start.go]: We don't need to since the cleanup is happening synchronously, except I suppose as a small backoff. But I don't think it's necessary.

File: daemon/start.go (hunk 8 above)
Comment (cpuguy83) [reply daemon/start.go]: Because we should figure out a way to plumb a proper context through here, otherwise the context is useless and is just being passed in to satisfy the argument list.

File: plugin/executor/containerd/containerd.go (hunk 1 above)
Comment (cpuguy83) [reply plugin/executor/containerd/containerd.go]: I think I read this from my phone before... `plugin task` works.

File: daemon/start.go (hunk 9 above)
Comment (cpuguy83) [reply daemon/start.go]: Actually, this is the wrong place. This should be handled on restore.... will post a fix.

File: daemon/start.go (hunk 8 above)
Comment (anusha-ragunathan) [reply daemon/start.go]: I've seen a lot of code where `context.Background` is passed simply to satisfy the argument. Glad to see TODO being used.

**Comment by cpuguy83** [comment]
I removed the stale container fix and will work on it for a separate PR to not block this.

File: plugin/manager_linux.go [hunk 10]
```diff
@@ -132,26 +131,32 @@ func (pm *Manager) restore(p *v2.Plugin) error {
 		return pm.pluginPostStart(p, c)
 	}
 
+	if alive {
+		// TODO(@cpuguy83): Should we always just re-attach to the running plugin instead of doing this?
```
Comment (anusha-ragunathan) [inline plugin/manager_linux.go]: remove this

File: plugin/manager.go [hunk 11]
```diff
@@ -143,12 +143,15 @@ func (pm *Manager) HandleExitEvent(id string) error {
...
-	os.RemoveAll(filepath.Join(pm.config.ExecRoot, id))
+	if err := os.RemoveAll(filepath.Join(pm.config.ExecRoot, id)); err != nil && !os.IsNotExist(err) {
+		logrus.WithError(err).WithField("id", id).Error("Could not remove plugin bundle dir")
+	}
 
 	pm.mu.RLock()
 	c := pm.cMap[p]
 	if c.exitChan != nil {
 		close(c.exitChan)
+		c.exitChan = nil
```
Comment (mlaventure) [inline plugin/manager.go]: could you explain why is this needed?

Comment (cpuguy83) [reply plugin/manager.go]: Because of the double-events from containerd we are seeing due to containerd/containerd#2299

Comment (cpuguy83) [reply plugin/manager.go]: It ends up causing a double close.

File: plugin/manager_linux.go (hunk 10 above)
Comment (cpuguy83) [reply plugin/manager_linux.go]: Why? This is a legitimate todo item.

Comment (anusha-ragunathan) [reply plugin/manager_linux.go]: Hm. Plugins should be best-effort consistent with container behavior. In that respect, I dont think we should try re-attach.

**Comment by cpuguy83** [comment]
Containers do what they do for historical reasons. But we always restart
plugins. Unless we have a really good reason why we shouldn't re-attach (in
which case that should be fixed) then I don't see why we should continue on
with the legacy from container behavior.

**Comment by anusha-ragunathan** [comment]
Then, lets decide and do the right thing in this PR. If we dont care about behavior consistency with containers, I agree with re-attaching. Lets do that in this PR, rather than have a follow up, only to change the behavior.

**Comment by cpuguy83** [comment]
I would prefer to fix the problem first, then we can discuss the behavior
change.

**Comment by thaJeztah** [comment]
Needs a rebase :cry:

**Comment by thaJeztah** [comment]
ping @anusha-ragunathan PTAL

**Comment by anusha-ragunathan** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @mlaventure good to go?

**Comment by jose-bigio** [comment]
@mlaventure does this look good to you?

**Comment by kolyshkin** [comment]
I'll take a look

**Comment by kolyshkin** [comment]
Please close this one in favor of https://github.com/moby/moby/pull/37234

**Comment by thaJeztah** [comment]
Carried in https://github.com/moby/moby/pull/37234
